pkg/ratelimit/limitware/middleware.go

```diff
@@ -0,0 +1,202 @@
+// This function returns an error if either the Store or KeyFunc are nil.
+func NewMiddleware(ctx context.Context, s limiter.Store, f httplimit.KeyFunc) (*Middleware, error) {
+	if s == nil {
+		return nil, fmt.Errorf("store cannot be nil")
+	}
+
+	if f == nil {
+		return nil, fmt.Errorf("key function cannot be nil")
+	}
+
+	logger := logging.FromContext(ctx).Named("ratelimit")
+
+	rc := stats.Int64(MetricPrefix+"/request", "requests seen by middleware", stats.UnitDimensionless)
+	if err := view.Register(&view.View{
+		Name:        MetricPrefix + "/request_count",
+		Measure:     rc,
+		Aggregation: view.Count(),
+		TagKeys:     []tag.Key{},
+	}); err != nil {
+		return nil, fmt.Errorf("stat view registration failure: %w", err)
+	}
+
+	ke := stats.Int64(MetricPrefix+"/key_errors", "errors seen from key function", stats.UnitDimensionless)
+	if err := view.Register(&view.View{
+		Name:        MetricPrefix + "/key_errors_count",
+		Measure:     ke,
+		Aggregation: view.Count(),
+		TagKeys:     []tag.Key{},
+	}); err != nil {
+		return nil, fmt.Errorf("stat view registration failure: %w", err)
+	}
+
+	return &Middleware{
+		store:      s,
+		keyFunc:    f,
+		reqCounter: rc,
+		keyErrors:  ke,
+		logger:     logger,
+	}, nil
+}
+
+// Handle returns the HTTP handler as a middleware. This handler calls Take() on
+// the store and sets the common rate limiting headers. If the take is
+// successful, the remaining middleware is called. If take is unsuccessful, the
+// middleware chain is halted and the function renders a 429 to the caller with
+// metadata about when it's safe to retry.
+func (m *Middleware) Handle(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Call the key function - if this fails, it's an internal server error.
+		key, err := m.keyFunc(r)
+		if err != nil {
+			m.logger.Errorw("could not call key function", "error", err)
+			stats.Record(r.Context(), m.keyErrors.M(1))
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			return
+		}
+
+		// Take from the store.
+		limit, remaining, reset, ok := m.store.Take(key)
+		resetTime := time.Unix(0, int64(reset)).UTC().Format(time.RFC1123)
+
+		// Set headers (we do this regardless of whether the request is permitted).
+		w.Header().Set(httplimit.HeaderRateLimitLimit, strconv.FormatUint(limit, 10))
+		w.Header().Set(httplimit.HeaderRateLimitRemaining, strconv.FormatUint(remaining, 10))
+		w.Header().Set(httplimit.HeaderRateLimitReset, resetTime)
+
+		// Record request status
+		ctx, err := tag.New(r.Context(), tag.Insert(OkTag, fmt.Sprintf("%v", ok)))
+		if err != nil {
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			return
+		}
+		stats.Record(ctx, m.reqCounter.M(1))
+
+		// Fail if there were no tokens remaining.
+		if !ok {
+			w.Header().Set(httplimit.HeaderRetryAfter, resetTime)
```

Comment:
Should we emit a metric here too? It'd be nice to be able to look at a dashboard and see the total number of currently rate limited requests.